source/rust_verify/src/singular.rs

```diff
@@ -339,67 +336,66 @@ pub fn check_singular_valid(
     let mut singular_process = singular_manager.launch();
-    let res = singular_process.send_commands(query.as_bytes().to_vec());
-    if (res.len() == 1) && (res[0] == "0") {
-        ValidityResult::Valid
-    } else if res[0].contains("?") {
-        ValidityResult::UnexpectedOutput(String::from(format!(
-            "{} \ngenerated singular query: {}",
-            res[0].as_str(),
-            query
-        )))
-    } else {
-        ValidityResult::Invalid(
-            None,
-            air::messages::error(
-                format!(
-                    "postcondition not satisfied: Ensures polynomial failed to be reduced to zero, reduced polynomial is {}\n generated singular query: {} ",
-                    res[0].as_str(),
-                    query
-                ),
-                &ens.1.spans[0],
-            ),
-        )
+
+    for stmt in &*stmts {
+        if let air::ast::StmtX::Assert(err, expr) = &**stmt {
+            assert_eq!(err.labels.len(), 1);
+            let span = &err.spans[0];
+            let label = &err.labels[0].note;
+            if label.contains("require") {
+                if let Err(info) = encoder.encode_requires_poly(expr) {
+                    return ValidityResult::Invalid(None, err.clone().secondary_label(span, info));
+                }
+            } else if label.contains("ensure") {
+                match encoder.encode_ensures_poly(expr) {
+                    Err(info) => {
+                        return ValidityResult::Invalid(
+                            None,
+                            err.clone().secondary_label(span, info),
+                        );
+                    }
+                    Ok(query) => {
+                        air::singular_manager::log_singular(context, &query, func_span);
+                        let res = singular_process.send_commands(query.as_bytes().to_vec());
+                        if (res.len() == 1) && (res[0] == "0") {
+                            continue;
+                        } else if (res.len() == 2) && (res[1] == "0") {
+                            assert!(res[0].contains("// ** redefining"));
+                            continue;
+                        } else if res[0].contains("?") {
+                            return ValidityResult::UnexpectedOutput(String::from(format!(
+                                "{} \ngenerated singular query: {}",
+                                res[0].as_str(),
+                                query
+                            )));
+                        } else {
+                            return ValidityResult::Invalid(
+                                None,
+                                err.clone().secondary_label(
+                                    span,
+                                    format!("Singular returned:\n {}", res.join("\n")),
+                                ),
+                            );
+                        }
```

Comment:
Can you please add a comment explaining how the output you are matching against indicates the outcome? (Why does "redefining" still allow for success? Why does the question mark indicate failure?)

Reply:
I added some comments on those.